ncm-dpmlfc/src/main/perl/dpmlfc/RuleBasedEditor.pm

```diff
@@ -0,0 +1,817 @@
+
+Arguments :
+    keyword : line keyword
+    value : keyword value (can be empty)
+    line_fmt : line format (see LINE_FORMAT_xxx constants)
+
+=cut
+
+sub _formatConfigLine {
+  my $function_name = "_formatConfigLine";
+  my ($self, $keyword, $value, $line_fmt) = @_;
+
+  unless ( $keyword ) {
+    $self->error("$function_name: 'keyword' argument missing (internal error)");
+    return 1;
+  }
+  unless ( defined($value) ) {
+    $self->error("$function_name: 'value' argument missing (internal error)");
+    return 1;
+  }
+  unless ( defined($line_fmt) ) {
+    $self->error("$function_name: 'line_fmt' argument missing (internal error)");
+    return 1;
+  }
+
+  my $config_line = "";
+
+  if ( $line_fmt == LINE_FORMAT_PARAM ) {
+    $config_line = "$keyword=$value";
+  } elsif ( $line_fmt == LINE_FORMAT_ENVVAR ) {
+    $config_line = "export $keyword=$value";
+  } elsif ( $line_fmt == LINE_FORMAT_XRDCFG_SETENV ) {
+    $config_line = "setenv $keyword = $value";
+  } elsif ( $line_fmt == LINE_FORMAT_XRDCFG_SET ) {
+    $config_line = "set $keyword = $value";
+  } elsif ( $line_fmt == LINE_FORMAT_XRDCFG ) {
+    $config_line = $keyword;
+    $config_line .= " $value" if $value;
+    # In trust (shift.conf) format, there should be only one blank between
+    # tokens and no trailing spaces.
+    $config_line =~ s/\s\s+/ /g;
+    $config_line =~ s/\s+$//;
+  } else {
+    $self->error("$function_name: invalid line format ($line_fmt). Internal inconsistency.");
+  }
+
+  $self->debug(2,"$function_name: Configuration line : >>$config_line<<");
+  return $config_line;
+}
+
+
+=pod
+
+=item _buildLinePattern
+
+This function builds a pattern that will match an existing configuration line for
+the configuration parameter specified. The pattern built takes into account the line format.
+Every whitespace in the pattern (configuration parameter) are replaced by \s+.
+If the line format is LINE_FORMAT_XRDCFG, no whitespace is
+imposed at the end of the pattern, as these format can be used to write a configuration
+directive as a keyword with no value.
+
+Arguments :
+    config_param: parameter to update
+    line_fmt: line format (see LINE_FORMAT_xxx constants)
+    config_value: when defined, make it part of the pattern (used when multiple lines
+                  with the same keyword are allowed)
+
+=cut
+
+sub _buildLinePattern {
+  my $function_name = "_buildLinePattern";
+  my ($self, $config_param, $line_fmt, $config_value) = @_;
+
+  unless ( $config_param ) {
+    $self->error("$function_name: 'config_param' argument missing (internal error)");
+    return undef;
+  }
+  unless ( defined($line_fmt) ) {
+    $self->error("$function_name: 'line_fmt' argument missing (internal error)");
+    return undef;
+  }
+  if ( defined($config_value ) ) {
+    $self->debug(2,"$function_name: configuration value '$config_value' will be added to the pattern");
+    $config_value =~ s/\\/\\\\/g;
+    $config_value =~ s/([\-\+\?\.\*\[\]()\^\$])/\\$1/g;
```

Comment:
For me (and based on my test), '\g1` can be used only inside the regexp, no on the substitution side....